From 9d37a84c98122ac046892d41d8419105ada31df8 Mon Sep 17 00:00:00 2001
From: Madhurkiran Harikrishnan <madhurkiran.harikrishnan@xilinx.com>
Date: Wed, 11 Dec 2019 15:34:06 -0800
Subject: [PATCH 0686/1851] drm: xlnx: zynqmp_disp: Perform additional checks
 for plane update

commit a41bdf237ed26c81626b890089a628a9ebc01b67 from
https://github.com/Xilinx/linux-xlnx.git

Relying on the checks for updating the plane on fb info alone may not be
sufficient as this might leave out corner case involving application
trying to show the same fb with different co-ordinates. Therefore,
perform additional checks to verify the co-ordinates(crtc_x, crtc_y,
src_x, src_y, etc.) of the old and the new plane states before bailing
out without updating the plane.

Signed-off-by: Madhurkiran Harikrishnan <madhurkiran.harikrishnan@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_disp.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_disp.c b/drivers/gpu/drm/xlnx/zynqmp_disp.c
index 6d643d12b626..103079fcb340 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_disp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_disp.c
@@ -2677,7 +2677,15 @@ zynqmp_disp_plane_atomic_update(struct drm_plane *plane,
 	if (!plane->state->crtc || !plane->state->fb)
 		return;
 
-	if (plane->state->fb == old_state->fb)
+	if (plane->state->fb == old_state->fb &&
+	    plane->state->crtc_x == old_state->crtc_x &&
+	    plane->state->crtc_y == old_state->crtc_y &&
+	    plane->state->crtc_w == old_state->crtc_w &&
+	    plane->state->crtc_h == old_state->crtc_h &&
+	    plane->state->src_x == old_state->src_x &&
+	    plane->state->src_y == old_state->src_y &&
+	    plane->state->src_w == old_state->src_w &&
+	    plane->state->src_h == old_state->src_h)
 		return;
 
 	if (old_state->fb &&
-- 
2.31.1

